[API Proposal]: Create array from array type

AlexRadch opened the pull request:
For https://github.com/dotnet/runtime/issues/76478

I added `InternalCreateFromArrayType` method. `InternalCreateFromArrayType` method is only correctly implemented for the `NativeAot` platform.

**At `CoreCLR` and `Mono` platforms it needs to be rewritten.** At `CoreCLR` and `Mono` it calls `InternalCreate` method which slows it down.

> The implementer is encouraged to consider whether all of these overloads are actually required, as the suggestion that the first overload will be in excess of 99% of all calls seems correct.

I decided to implement only the next overloads:

```cs
namespace System;

public partial class Array
{
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params int[] lengths);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int[] lengths, int[] lowerBounds);
}
```

And I have not implemented the following overloads:

```cs
namespace System;

public partial class Array
{
        // It is not used in the current code and overload with `params int[] lengths` can be used instead
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2); 

        // It is not used in the current code and overload with `params int[] lengths` can be used instead
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2, int length3);

        // Array does not support lengths more than `int.Max`. Array always uses `int` for lengths, not `long`
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params long[] lengths);
}
```

AlexRadch committed:
Create array from array type

huoyaoyuan commented:
> And I have not implemented the following overloads:

They are approved for consistency with current overloads.

> It is not used in the current code and overload with `params int[] lengths` can be used instead

It should be somehow useful in third party code. Params array requires allocation, so we typically define overloads for two or three parameters.

If you have concerns about these overloads, you can provide feedback in original issue.

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Array.cs:
```diff
@@ -207,6 +207,74 @@ public static Array CreateInstance(Type elementType, params long[] lengths)
             return CreateInstance(elementType, intLengths);
         }
 
+        public static unsafe Array CreateInstanceFromArrayType(Type arrayType, int length)
```
I added `DynamicallyAccessedMembers` in a first try, and caused other places accessing arrays starting to report trimmer warning. This is an actual blocker to resolve.

jkotas replied:
That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be.

huoyaoyuan replied:
The actual problem is quite complicated and I can't explain its detail without my machine. In short, something in `System.Xml` is passing array type to method annotated with `DynamicallyAccessedMembers.All`, and thus the static members are considered access, then the `DynamicallyAccessedMembers.Constructors` triggers warning. It may be resolved in some trimmer change, I haven't confirmed yet.

AlexRadch replied:
> That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be.

@jkotas  I understand that. What can I use instead of `InternalCreate`? What method does the compiler use?

AlexRadch replied:
I think it is impossible to use pathed `arrayType` in all cases 😥. I think in 99,9% of cases we can use `arrayType` to create an array but sometimes not. For example `arrayType` can be a type for a one-dimensional array and the user want to create a multi-dimensional array. So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

MichalStrehovsky replied:
> I think it is impossible to use pathed `arrayType` in all cases 😥. I think in 99,9% of cases we can use `arrayType` to create an array but sometimes not. For example `arrayType` can be a type for a one-dimensional array and the user want to create a multi-dimensional array. So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

Somehow Github made this comment appear twice but this is my response to the other instance: https://github.com/dotnet/runtime/pull/88620#discussion_r1259420229. We don't need that behavior. If type doesn't match dimensions, it's an error.

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Array.cs:
```diff
@@ -207,6 +207,74 @@ public static Array CreateInstance(Type elementType, params long[] lengths)
+            if (!arrayType.IsArray)
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Arg_MustBeArrayType, ExceptionArgument.arrayType);
+
+            RuntimeType t = (arrayType.GetElementType()!.UnderlyingSystemType as RuntimeType)!;
+            Debug.Assert(t is not null, $"CreateInstanceFromArrayType can not get underlying system type for \"{arrayType}\"");
```
The `t` variable can be defined as nullable to save a `!`. I think we are allowing `var` here.

AlexRadch replied:
`t` can not be null here so I use `RuntimeType`, not `RuntimeType?`.

AlexRadch commented:
> They are approved for consistency with current overloads.

They wrote

> The implementer is encouraged to consider whether all of these overloads are actually required, as the suggestion that the first overload will be in excess of 99% of all calls seems correct.

AlexRadch commented:
>> It is not used in the current code and overload with params int[] lengths can be used instead

> It should be somehow useful in third party code. Params array requires allocation, so we typically define overloads for two or three parameters.

Methods with two or three parameters will allocate an array inside to call `InternalCreate` method. So it does not reduce allocations.

huoyaoyuan commented:
> Methods with two or three parameters will allocate an array inside to call `InternalCreate` method.

They shouldn't. `CreateInstance` uses `stackalloc`.

AlexRadch commented:
> They shouldn't. `CreateInstance` uses `stackalloc`.

You are right, `CreateInstance` uses `stackalloc`. These methods are so rarely used .Net library does not use them at all. Can the compiler also use stack allocation? I think they are actually NOT required.

huoyaoyuan commented:
> Can the compiler also use stack allocation?

That's params Span<T> designed for (https://github.com/dotnet/csharplang/issues/1757). The prerequisite work has started, and I'd expect that feature to come in C# 12 or 13.

AlexRadch commented:
> That's params Span designed for ([dotnet/csharplang#1757](https://github.com/dotnet/csharplang/issues/1757)). The prerequisite work has started, and I'd expect that feature to come in C# 12 or 13.

So method overloads with two and three params are NOT required. Later we will add `params Span` overload and it will not allocate memory.

MichalStrehovsky commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeInstantiator.cs:
```diff
@@ -144,8 +144,7 @@ public static Attribute Instantiate(this CustomAttributeData cad)
                     return null;
-                Type? elementType = argumentType.GetElementType();
-                Array array = Array.CreateInstance(elementType, typedElements.Count);
+                Array array = Array.CreateInstanceFromArrayType(argumentType, typedElements.Count);
```
I think the `UnconditionalSuppressMessage` attribute at the top of this method can now be deleted - it was suppressing a warning caused by the Array.CreateInstance. I don't see anything else in this method that would generate this warning.

AlexRadch committed:
Code review

MichalStrehovsky commented on src/libraries/System.Private.CoreLib/src/System/Array.cs:
```diff
@@ -207,6 +207,77 @@ public static Array CreateInstance(Type elementType, params long[] lengths)
             return CreateInstance(elementType, intLengths);
         }
 
+        [RequiresDynamicCode("The code for an array of the specified type might not be available.")]
```
The new APIs will not be annotated RequiresDynamicCode. These APIs being AOT safe was the explicit purpose for adding them (see "and requires suppressing warnings about the array type potentially not being available with Aot." in the issue).

The implementation currently generates a warning but that because of what jkotas said below: "That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be". We don't want the implementation to go back to array. It negates the whole point of the new API.

huoyaoyuan replied:
Does it need `DynamicallyAccessedMembers.PublicConstructors` on the array type? That would trigger the problem I described.

AlexRadch replied:
> The implementation currently generates a warning but that because of what jkotas said below: "That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be". We don't want the implementation to go back to array. It negates the whole point of the new API.

I think it is impossible in all cases 😥. I think in 99,9% of cases we can use `arrayType` to create an array but sometimes not. For example `arrayType` can be a type for a one-dimensional array and the user want to create a multi-dimensional array. So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

MichalStrehovsky replied:
> So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

I think the method should throw an exception if the dimensions don't match the array type.

It's the same as if we had a `ConstructorInfo` for a method that takes 1 parameter, but call it with two parameters. The BCL could in theory go looking for a different constructor on the type that matches the passed parameters. But it doesn't do that - it's an invalid use.

AlexRadch committed:
InternalCreateFromArrayType

AlexRadch committed:
GC

AlexRadch committed:
GC

MichalStrehovsky commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -4235,4 +4235,7 @@
   </data>
-</root>
+  <data name="Arg_MustBeArrayType" xml:space="preserve">
+    <value>Type must be a array type.</value>
```
We have `Argument_HasToBeArrayClass` that is used in similar situations - I think we can just reuse that string.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs:
```diff
@@ -18,6 +17,28 @@ public abstract partial class Array : ICloneable, IList, IStructuralComparable,
 
+        private static unsafe Array InternalCreateFromArrayType(Type arrayType, int rank, int* pLengths, int* pLowerBounds)
+        {
+            if (rank == 1 && !ContainsLowerBounds(rank, pLowerBounds))
```
```suggestion
            if (rank == 1 && pLowerBounds[0] == 0)
```
And delete ContainsLowerBounds

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs:
```diff
@@ -18,6 +17,28 @@ public abstract partial class Array : ICloneable, IList, IStructuralComparable,
+        {
+            if (rank == 1 && !ContainsLowerBounds(rank, pLowerBounds))
+            {
+                return GC.AllocateNewArray(arrayType.TypeHandle.Value, pLengths[0], GC.GC_ALLOC_FLAGS.GC_ALLOC_NO_FLAGS);
```
`GC.AllocateNewArray` is designed for allocating the special arrays. You may want to measure how slow it is for allocating ordinary small arrays.

We may want to have a new runtime calls for this that will handle the multi-dim case too

AlexRadch replied:
I got the next benchmark scores:
```
|             Type |                Method |   Count |             Mean |          Error |         StdDev |           Median | Ratio | RatioSD |     Gen0 |     Gen1 |     Gen2 | Allocated | Alloc Ratio |
|----------------- |---------------------- |-------- |-----------------:|---------------:|---------------:|-----------------:|------:|--------:|---------:|---------:|---------:|----------:|------------:|
|    BenchArrayInt |                Create |       0 |         8.531 ns |      0.0618 ns |      0.0548 ns |         8.537 ns |  1.00 |    0.00 |   0.0057 |        - |        - |      24 B |        1.00 |
|    BenchArrayInt |              Allocate |       0 |        45.232 ns |      0.9173 ns |      1.2245 ns |        44.796 ns |  5.36 |    0.18 |   0.0057 |        - |        - |      24 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |       0 |         8.971 ns |      0.0736 ns |      0.0652 ns |         8.985 ns |  1.05 |    0.01 |   0.0057 |        - |        - |      24 B |        1.00 |
| BenchArrayString |                Create |       0 |         8.193 ns |      0.1971 ns |      0.3554 ns |         8.040 ns |  0.96 |    0.04 |   0.0057 |        - |        - |      24 B |        1.00 |
| BenchArrayString |              Allocate |       0 |        46.420 ns |      0.9399 ns |      1.4353 ns |        46.031 ns |  5.51 |    0.19 |   0.0057 |        - |        - |      24 B |        1.00 |
| BenchArrayString | AllocateUninitialized |       0 |         7.978 ns |      0.1903 ns |      0.1869 ns |         7.957 ns |  0.94 |    0.02 |   0.0057 |        - |        - |      24 B |        1.00 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |       1 |        10.022 ns |      0.0600 ns |      0.0532 ns |        10.007 ns |  1.00 |    0.00 |   0.0076 |        - |        - |      32 B |        1.00 |
|    BenchArrayInt |              Allocate |       1 |        45.933 ns |      0.6901 ns |      0.6117 ns |        45.977 ns |  4.58 |    0.06 |   0.0076 |        - |        - |      32 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |       1 |        10.480 ns |      0.1064 ns |      0.0943 ns |        10.459 ns |  1.05 |    0.01 |   0.0076 |        - |        - |      32 B |        1.00 |
| BenchArrayString |                Create |       1 |         9.512 ns |      0.2020 ns |      0.1791 ns |         9.442 ns |  0.95 |    0.02 |   0.0076 |        - |        - |      32 B |        1.00 |
| BenchArrayString |              Allocate |       1 |        47.115 ns |      0.4613 ns |      0.4089 ns |        47.239 ns |  4.70 |    0.04 |   0.0076 |        - |        - |      32 B |        1.00 |
| BenchArrayString | AllocateUninitialized |       1 |         9.621 ns |      0.1906 ns |      0.3533 ns |         9.533 ns |  0.99 |    0.04 |   0.0076 |        - |        - |      32 B |        1.00 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |       5 |        13.273 ns |      0.2988 ns |      0.3197 ns |        13.212 ns |  1.00 |    0.00 |   0.0115 |        - |        - |      48 B |        1.00 |
|    BenchArrayInt |              Allocate |       5 |        48.570 ns |      0.7197 ns |      0.6380 ns |        48.606 ns |  3.64 |    0.10 |   0.0114 |        - |        - |      48 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |       5 |        13.704 ns |      0.1898 ns |      0.1776 ns |        13.657 ns |  1.03 |    0.02 |   0.0115 |        - |        - |      48 B |        1.00 |
| BenchArrayString |                Create |       5 |        15.731 ns |      0.1732 ns |      0.1353 ns |        15.740 ns |  1.17 |    0.03 |   0.0153 |        - |        - |      64 B |        1.33 |
| BenchArrayString |              Allocate |       5 |        53.037 ns |      0.3743 ns |      0.3318 ns |        53.119 ns |  3.97 |    0.10 |   0.0153 |        - |        - |      64 B |        1.33 |
| BenchArrayString | AllocateUninitialized |       5 |        15.644 ns |      0.1767 ns |      0.1653 ns |        15.627 ns |  1.17 |    0.03 |   0.0153 |        - |        - |      64 B |        1.33 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |      10 |        15.196 ns |      0.1711 ns |      0.1429 ns |        15.183 ns |  1.00 |    0.00 |   0.0153 |        - |        - |      64 B |        1.00 |
|    BenchArrayInt |              Allocate |      10 |        51.142 ns |      0.4151 ns |      0.3467 ns |        51.112 ns |  3.37 |    0.04 |   0.0153 |        - |        - |      64 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |      10 |        16.782 ns |      0.2404 ns |      0.2249 ns |        16.719 ns |  1.11 |    0.02 |   0.0153 |        - |        - |      64 B |        1.00 |
| BenchArrayString |                Create |      10 |        23.496 ns |      0.3257 ns |      0.2720 ns |        23.519 ns |  1.55 |    0.02 |   0.0249 |        - |        - |     104 B |        1.62 |
| BenchArrayString |              Allocate |      10 |        63.847 ns |      0.8255 ns |      0.6445 ns |        64.097 ns |  4.20 |    0.04 |   0.0248 |        - |        - |     104 B |        1.62 |
| BenchArrayString | AllocateUninitialized |      10 |        23.543 ns |      0.2681 ns |      0.2377 ns |        23.540 ns |  1.55 |    0.02 |   0.0249 |        - |        - |     104 B |        1.62 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |     100 |        83.160 ns |      1.6437 ns |      1.7587 ns |        82.706 ns |  1.00 |    0.00 |   0.1013 |        - |        - |     424 B |        1.00 |
|    BenchArrayInt |              Allocate |     100 |       120.106 ns |      1.9915 ns |      1.8629 ns |       119.715 ns |  1.44 |    0.05 |   0.1013 |        - |        - |     424 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |     100 |        84.808 ns |      0.6549 ns |      0.5113 ns |        84.780 ns |  1.01 |    0.02 |   0.1013 |        - |        - |     424 B |        1.00 |
| BenchArrayString |                Create |     100 |       160.460 ns |      1.8637 ns |      1.7433 ns |       160.166 ns |  1.93 |    0.05 |   0.1969 |        - |        - |     824 B |        1.94 |
| BenchArrayString |              Allocate |     100 |       196.155 ns |      2.7282 ns |      2.4185 ns |       195.607 ns |  2.35 |    0.04 |   0.1969 |        - |        - |     824 B |        1.94 |
| BenchArrayString | AllocateUninitialized |     100 |       159.295 ns |      2.0602 ns |      1.7203 ns |       159.233 ns |  1.91 |    0.04 |   0.1969 |        - |        - |     824 B |        1.94 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |    1000 |       777.217 ns |     12.3406 ns |     10.9396 ns |       775.492 ns |  1.00 |    0.00 |   0.9613 |        - |        - |    4024 B |        1.00 |
|    BenchArrayInt |              Allocate |    1000 |       789.251 ns |     10.2752 ns |      9.6114 ns |       787.657 ns |  1.02 |    0.01 |   0.9613 |        - |        - |    4024 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |    1000 |       187.433 ns |      3.7948 ns |      7.6657 ns |       184.953 ns |  0.25 |    0.01 |   0.9584 |        - |        - |    4024 B |        1.00 |
| BenchArrayString |                Create |    1000 |     1,554.818 ns |     15.1447 ns |     14.1664 ns |     1,559.290 ns |  2.00 |    0.04 |   1.9150 |        - |        - |    8024 B |        1.99 |
| BenchArrayString |              Allocate |    1000 |     1,562.925 ns |     13.4857 ns |     11.9547 ns |     1,563.005 ns |  2.01 |    0.03 |   1.9150 |        - |        - |    8024 B |        1.99 |
| BenchArrayString | AllocateUninitialized |    1000 |     1,562.850 ns |     19.5024 ns |     17.2883 ns |     1,564.989 ns |  2.01 |    0.03 |   1.9150 |        - |        - |    8024 B |        1.99 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |   10000 |     7,200.916 ns |     67.9913 ns |     63.5991 ns |     7,185.784 ns |  1.00 |    0.00 |   9.5215 |        - |        - |   40024 B |        1.00 |
|    BenchArrayInt |              Allocate |   10000 |     7,149.810 ns |     84.2155 ns |     78.7753 ns |     7,130.345 ns |  0.99 |    0.02 |   9.5215 |        - |        - |   40024 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |   10000 |       651.649 ns |      7.0545 ns |      5.8909 ns |       650.162 ns |  0.09 |    0.00 |   9.5234 |        - |        - |   40024 B |        1.00 |
| BenchArrayString |                Create |   10000 |    14,001.614 ns |    237.3965 ns |    198.2368 ns |    14,009.929 ns |  1.94 |    0.03 |  18.8599 |        - |        - |   80024 B |        2.00 |
| BenchArrayString |              Allocate |   10000 |    14,089.164 ns |    142.5671 ns |    119.0500 ns |    14,099.614 ns |  1.96 |    0.02 |  18.8599 |        - |        - |   80024 B |        2.00 |
| BenchArrayString | AllocateUninitialized |   10000 |    14,088.178 ns |     65.6579 ns |     58.2040 ns |    14,101.808 ns |  1.95 |    0.02 |  18.8599 |        - |        - |   80024 B |        2.00 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create |  100000 |    29,484.404 ns |    311.6528 ns |    382.7376 ns |    29,507.994 ns |  1.00 |    0.00 | 124.9695 | 124.9695 | 124.9695 |  400066 B |        1.00 |
|    BenchArrayInt |              Allocate |  100000 |    29,567.013 ns |    449.1566 ns |    420.1414 ns |    29,604.886 ns |  1.00 |    0.02 | 124.9695 | 124.9695 | 124.9695 |  400066 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized |  100000 |    26,492.538 ns |    292.9388 ns |    259.6827 ns |    26,564.154 ns |  0.90 |    0.02 | 124.9695 | 124.9695 | 124.9695 |  400066 B |        1.00 |
| BenchArrayString |                Create |  100000 |    54,173.601 ns |    672.7485 ns |    596.3741 ns |    54,227.924 ns |  1.84 |    0.04 | 249.9390 | 249.9390 | 249.9390 |  800107 B |        2.00 |
| BenchArrayString |              Allocate |  100000 |    54,339.854 ns |    967.7527 ns |    808.1174 ns |    54,227.414 ns |  1.85 |    0.03 | 249.9390 | 249.9390 | 249.9390 |  800107 B |        2.00 |
| BenchArrayString | AllocateUninitialized |  100000 |    53,836.765 ns |    441.4254 ns |    391.3121 ns |    53,730.786 ns |  1.83 |    0.03 | 249.9390 | 249.9390 | 249.9390 |  800107 B |        2.00 |
|                  |                       |         |                  |                |                |                  |       |         |          |          |          |           |             |
|    BenchArrayInt |                Create | 1000000 |   208,606.318 ns |  3,285.4765 ns |  3,073.2368 ns |   208,015.698 ns |  1.00 |    0.00 | 999.7559 | 999.7559 | 999.7559 | 4000356 B |        1.00 |
|    BenchArrayInt |              Allocate | 1000000 |   207,851.978 ns |  3,900.6592 ns |  3,830.9689 ns |   206,175.659 ns |  1.00 |    0.02 | 999.7559 | 999.7559 | 999.7559 | 4000357 B |        1.00 |
|    BenchArrayInt | AllocateUninitialized | 1000000 |   183,179.046 ns |  1,604.5172 ns |  1,500.8664 ns |   183,096.216 ns |  0.88 |    0.01 | 999.7559 | 999.7559 | 999.7559 | 4000329 B |        1.00 |
| BenchArrayString |                Create | 1000000 | 2,440,452.062 ns | 48,071.0001 ns | 64,173.4044 ns | 2,449,305.859 ns | 11.64 |    0.36 | 140.6250 | 140.6250 | 140.6250 | 8000068 B |        2.00 |
| BenchArrayString |              Allocate | 1000000 | 2,427,959.420 ns | 37,518.6029 ns | 31,329.7327 ns | 2,437,695.117 ns | 11.68 |    0.20 | 132.8125 | 132.8125 | 132.8125 | 8000066 B |        2.00 |
| BenchArrayString | AllocateUninitialized | 1000000 | 2,412,512.165 ns | 33,286.0854 ns | 29,507.2477 ns | 2,419,748.047 ns | 11.60 |    0.18 | 140.6250 | 140.6250 | 140.6250 | 8000068 B |        2.00 |

```

jkotas replied:
Thanks, these benchmark shows that `GC.Allocate` has high overhead for small sizes, but they do not show where the new API is going to land exactly.

I would be interesting to benchmark `new int[1]` vs. `Array.CreateInstanceFromArrayType(typeof(int[]), 1)` vs. `Array.CreateInstance(typeof(int), 1)`. We want `Array.CreateInstanceFromArrayType` performance to be as close as possible to `new int[1]`.

AlexRadch committed:
code review

AlexRadch committed:
code review Exceptions

jkotas commented on src/libraries/System.Runtime/tests/System/ArrayTests.cs:
```diff
@@ -1923,6 +1949,13 @@ public void CreateInstance_NotSupportedType_ThrowsNotSupportedException(Type ele
             Assert.Throws<NotSupportedException>(() => Array.CreateInstance(elementType, new long[1]));
             Assert.Throws<NotSupportedException>(() => Array.CreateInstance(elementType, new int[1], new int[1]));
+
+            Assert.Throws<ArgumentException>("arrayType", () => Array.CreateInstanceFromArrayType(elementType, 0));
```
```suggestion
            Assert.Throws<ArgumentException>("arrayType", () => Array.CreateInstanceFromArrayType(elementType.MakeArrayType(), 0));
```
These tests need to create the array type from the elementType.

I expect that some of them going to fail since the implementation does not have proper error handling. It is a bug in the implementation that should be fixed.

jkotas commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:
```diff
@@ -572,8 +572,7 @@ private void WriteMembers(Member[] members, UnknownNodeAction elementElseAction,
                 {
-                    Type elementType = collectionType.GetElementType()!;
-                    a = Array.CreateInstance(elementType, collectionMember.Count);
+                    a = Array.CreateInstanceFromArrayType(collectionType, collectionMember.Count);
```
You can also fix the TODO here: Replace with https://github.com/dotnet/runtime/issues/76478

It should be fixed by creating the array using `Array.CreateInstanceFromArrayType` and then calling `ArrayList.CopyTo` to fill it in.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into Array.CreateInstanceFromArrayType

adamsitnik committed:
code review:

- add missing comments for new public APIs
- use the new API in DataTypeImplementation (and solve a TODO)
- fix typo
- remove code that was commented out
- improve test name

adamsitnik reviewed (approved):
Since this PR was stale for a long time and it seems that @AlexRadch has been inactive for a while, I decided to simply push fixes rather than leaving comments: https://github.com/dotnet/runtime/pull/88620/commits/f32e6e5c720b099d2302500d33443ce03197523e @jkotas PTAL

@AlexRadch thank you for your contribution!

jkotas commented on src/libraries/System.Private.Xml/src/System/Xml/Schema/DataTypeImplementation.cs:
```diff
@@ -1143,11 +1143,12 @@ internal override RestrictionFlags ValidRestrictionFlags
-            // TODO: Replace with https://github.com/dotnet/runtime/issues/76478 once available
-            [UnconditionalSuppressMessage("AotAnalysis", "IL3050:AotUnfriendlyApi",
-                Justification = "Array type is always present as it is passed in as a parameter.")]
             static Array ToArray(ArrayList values, Type arrayType)
```
This does not need to be in a separate method now that it does not have UnconditionalSuppressMessage annotation
```
if (_itemType.Variety == XmlSchemaDatatypeVariety.Union)
{
...
    arrayType = typeof(XmlAtomicValue[]);
}
else
{
....
    arrayType = _itemType.ListValueType;
}
if (values.Count < _minListSize)
{
    return new XmlSchemaException(SR.Sch_EmptyAttributeValue, string.Empty);
}
Array array = Array.CreateInstanceFromArrayType(arrayType, values.Count);
values.CopyTo(result);
```

adamsitnik committed:
address code  review feedback: simplify the code

jkotas commented on src/libraries/System.Private.Xml/src/System/Xml/Schema/DataTypeImplementation.cs:
```diff
@@ -1126,8 +1126,12 @@ internal override RestrictionFlags ValidRestrictionFlags
+                arrayType = _itemType.ListValueType;
             }
+
+            Array array = Array.CreateInstanceFromArrayType(arrayType, values.Count);
```
Nit: This can be after `if (values.Count < _minListSize)` check. We do not need to create the array when returning an exception.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs:
```diff
@@ -18,6 +17,12 @@ public abstract partial class Array : ICloneable, IList, IStructuralComparable,
 
+        private static unsafe Array InternalCreateFromArrayType(Type arrayType, int rank, int* pLengths, int* pLowerBounds)
+        {
+            return rank == 1 && (pLowerBounds == null || pLowerBounds[0] == 0)
```
This is leaving a lot of perf on the table. `CreateInstanceFromArrayType` should be faster than `CreateInstance` since `CreateInstanceFromArrayType` needs to do less work, but it is the case for this implementation. 

You can micro-benchmark
```
    int[] arg = new int[] { 1, 1 };
    GC.KeepAlive(Array.CreateInstance(typeof(int), arg));
```
vs.
```
    int[] arg = new int[] { 1, 1 };
    GC.KeepAlive(Array.CreateInstanceFromArrayType(typeof(int[,]), arg));
```
`CreateInstanceFromArrayType` is 10% slower on my machine.

jkotas replied:
`CreateInstanceFromArrayType` is about 1.4x faster compared to `CreateInstance` now (CoreCLR, Windows, x64).

adamsitnik commented:
> The error handling for invalid types is not quite right. 
> I think it should fail with the same exception as Array.CreateInstance.

I see. But how can I do that without calling `type.GetElementType()`?

https://github.com/dotnet/runtime/blob/9ad24aec018f8dbb4e5d5f67390f595930b5ea18/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.NativeAot.cs#L89-L101

jkotas commented:
> I see. But how can I do that without calling type.GetElementType()?

I think you either need to call `type.GetElementType()` or do the error handling in the unmanaged code.

adamsitnik commented:
> I think you either need to call type.GetElementType() 

> I think you either need to call type.GetElementType() or do the error handling in the unmanaged code.

I did some quick testing and it seems that C# does not allow for creating an illegal array type via `typeof` syntax:

```cs
_ = typeof(void[]);
_ = typeof(Span<int>[]);
_ = typeof(List<>[]);
```

And `Type.MakeArrayType` seems to be the only way of creating such arrays.  @jkotas What is the point of allowing this method to do it? Perhaps it should just throw in such cases, so all types that return `true` for `IsArray` would be OK?

If throwing is not acceptable, then would moving this check to `MakeArrayType` and extending the type with some internal boolean flag field used by `Array` (they are in the same lib) be acceptable?

adamsitnik committed:
address code review feedback

adamsitnik committed:
supress new test on Mono, which seems to affect other tests as well, see #94086 for details

jkotas committed:
Merge branch 'main' into Array.CreateInstanceFromArrayType

jkotas committed:
Fix perf and corner-case behaviors

jkotas committed:
Merge branch 'main' into Array.CreateInstanceFromArrayType

jkotas committed:
Cleanup

jkotas committed:
Delete superfluous validation

jkotas committed:
Fix build break

jkotas committed:
Smaller diff

jkotas committed:
Fix build break

adamsitnik commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/ArrayTests.cs:
```diff
@@ -1980,12 +2049,33 @@ public void CreateInstance_LengthsEmpty_ThrowsArgumentException()
+
+            Assert.Equal(Array.CreateInstanceFromArrayType(variableBoundArrayType, 12).GetType(), typeof(object[]));
+            Assert.Equal(Array.CreateInstanceFromArrayType(variableBoundArrayType, [22]).GetType(), typeof(object[]));
+            Assert.Equal(Array.CreateInstanceFromArrayType(variableBoundArrayType, [33], [0]).GetType(), typeof(object[]));
```
build fix

```suggestion
            Assert.Equal(typeof(object[]), Array.CreateInstanceFromArrayType(variableBoundArrayType, [33], [0]).GetType());
```

adamsitnik commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/ArrayTests.cs:
```diff
@@ -1980,12 +2049,33 @@ public void CreateInstance_LengthsEmpty_ThrowsArgumentException()
+            Assert.NotEqual(variableBoundArrayType, typeof(object[]));
+
+            Assert.Equal(Array.CreateInstanceFromArrayType(variableBoundArrayType, 12).GetType(), typeof(object[]));
+            Assert.Equal(Array.CreateInstanceFromArrayType(variableBoundArrayType, [22]).GetType(), typeof(object[]));
```
```suggestion
            Assert.Equal(typeof(object[]), Array.CreateInstanceFromArrayType(variableBoundArrayType, [22]).GetType());
```

adamsitnik commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/ArrayTests.cs:
```diff
@@ -1980,12 +2049,33 @@ public void CreateInstance_LengthsEmpty_ThrowsArgumentException()
+            Type variableBoundArrayType = typeof(object).MakeArrayType(1);
+            Assert.NotEqual(variableBoundArrayType, typeof(object[]));
+
+            Assert.Equal(Array.CreateInstanceFromArrayType(variableBoundArrayType, 12).GetType(), typeof(object[]));
```
```suggestion
            Assert.Equal(typeof(object[]), Array.CreateInstanceFromArrayType(variableBoundArrayType, 12).GetType());
```

adamsitnik commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/ArrayTests.cs:
```diff
@@ -1980,12 +2049,33 @@ public void CreateInstance_LengthsEmpty_ThrowsArgumentException()
+        public void CreateInstanceFromTemplate_Rank1()
+        {
+            Type variableBoundArrayType = typeof(object).MakeArrayType(1);
+            Assert.NotEqual(variableBoundArrayType, typeof(object[]));
```
```suggestion
            Assert.NotEqual(typeof(object[]), variableBoundArrayType);
```

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Array.cs:
```diff
@@ -186,12 +174,7 @@ public static unsafe Array CreateInstance(Type elementType, int[] lengths, int[]
-            {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.lengths);
-            }
-            if (lengths.Length == 0)
-                ThrowHelper.ThrowArgumentException(ExceptionResource.Arg_NeedAtLeast1Rank);
```
Where exactly is the `lengths.Length == 0` check performed now?

jkotas replied:
It will be performed by the `int[]` overload that this forwards to.

adamsitnik commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.NativeAot.cs:
```diff
@@ -72,34 +72,68 @@ private static unsafe Array InternalCreate(RuntimeType elementType, int rank, in
-#pragma warning disable CA1859 // https://github.com/dotnet/roslyn-analyzers/issues/6451
-        private static void ValidateElementType(Type elementType)
+        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
+            Justification = "The compiler ensures that if we have a TypeHandle of a Rank-1 MdArray, we also generated the SzArray.")]
```
thank you for adding a detailed explanation 👍

adamsitnik commented on src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs:
```diff
@@ -15,9 +14,18 @@ namespace System
+        }
 
+        private static unsafe Array InternalCreateFromArrayType(RuntimeType type, int rank, int* pLengths, int* pLowerBounds)
+            => InternalCreate(type, -rank, pLengths, pLowerBounds);
```
subjective nit: using `-` to change the meaning of another argument is used here, but commented in the native code. I wonder whether we could add a short comment here as well (or have two type arguments: `elementType` and `arrayType` and just check for nulls)

jkotas replied:
I am replacing this trick with a self-describing `bool` argument.

adamsitnik reviewed (commented):
Thank you for your help @jkotas!

adamsitnik committed:
build fix

jkotas committed:
FB, fix tests

jkotas commented on src/coreclr/classlibnative/bcltype/arraynative.h:
```diff
@@ -66,6 +61,7 @@ class ArrayNative
 
 };
 
+extern "C" void QCALLTYPE Array_CreateInstance(QCall::TypeHandle pTypeHnd, INT32 rank, INT32* pLengths, INT32* pBounds, BOOL createFromArrayType, QCall::ObjectHandleOnStack retArray);
```
Converting FCALL to QCALL makes the existing CreateInstance up to 10% faster.

jkotas committed:
Add back workaround for Mono

jkotas committed:
Skip handling of void arrays for this PR

adamsitnik commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/ArrayTests.cs:
```diff
@@ -1950,10 +1946,7 @@ public void CreateInstance_NotSupportedType_ThrowsNotSupportedException(Type ele
-        [InlineData(typeof(void))]
         [InlineData(typeof(GenericClass<>))]
-        // not using any by-ref type here as MakeArrayType throws for them, same goes for Type.GetType("SomeByRef[]")	
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/94086", typeof(PlatformDetection), nameof(PlatformDetection.IsMonoRuntime))]
```
if we remove the link to the issue it will be harder to reproduce for the Mono Team

How about leaving it like this?

```cs
// [InlineData(typeof(void))] fails on Mono: https://github.com/dotnet/runtime/issues/94086
```

jkotas replied:
My plan is to disallow arrays of System.Void in a follow up PR that will fix the Mono issue as a side-effect: https://github.com/dotnet/runtime/pull/94835

adamsitnik replied:
@jkotas sounds like a much better plan!

jkotas committed:
Fix tests for native AOT

jkotas committed:
Revert unnecessary change

jkotas commented:
@adamsitnik This PR should be green now. Any other feedback?

adamsitnik reviewed (approved):
@jkotas LGTM, big thanks for your help!
